### ISSUE TITLE
zeroize: provide a Vec with zero-on-realloc?

### ISSUE THREAD

**Issue by FauxFaux** [iqlusioninc/crates#179]
Zeroing a `Vec` on `Drop` may still leave data on the heap, as internally `Vec` may realloc-and-copy the data to the new size.

Is it possible to provide a `Vec`-like type which zeros the old data during reallocation?

Custom allocators, or hooks in the (custom) global allocator, may be future possible alternatives.

I believe that a typical `realloc` implementation will not copy, but it is by no means guaranteed or even likely to be true in all cases.

Spotted by @cuviper.

**Comment by tarcieri** [comment]
Yep, I should probably call out `Vec` explicitly in the docs under the caveats section in regard to things which make copies. I have definitely ranted about it offline quite a bit, and wondered if providing `Zeroize` impls for these types, including `String`, is a good idea or not. As this crate is effectively "best effort", the intent is to zero out as many copies as possible with the assumption fewer copies is better, leaving avoiding making them in the first place as an exercise to the end user.

> Is it possible to provide a `Vec`-like type which zeros the old data during reallocation?

One option is to use [heapless::Vec](https://docs.rs/heapless/latest/heapless/struct.Vec.html) and [heapless::String](https://docs.rs/heapless/latest/heapless/struct.String.html). Otherwise I think functionality beyond the core zeroization behavior is best left to crates which consume zeroize and leverage it as a zeroing backend.

**Comment by tarcieri** [comment]
I recently created `secrecy`, which seems like an appropriate place for these sorts of types:

https://github.com/iqlusioninc/crates/tree/develop/secrecy

I'd like to add `SecretVec` and a `SecretString` (which is backed by `SecretVec`)
